lib/Dialect/Transforms/IsolateGroupOpsPass.cpp

```diff
@@ -29,45 +29,55 @@ namespace mlir::tcp {
 
 namespace {
 
-class IsolateGroups : public OpRewritePattern<tcp::GroupOp> {
+class IsolateGroups : public OpRewritePattern<GroupOp> {
 public:
-  using OpRewritePattern::OpRewritePattern;
+  using OpRewritePattern<GroupOp>::OpRewritePattern;
 
-  LogicalResult matchAndRewrite(tcp::GroupOp groupOp,
+  IsolateGroups(MLIRContext *context,
+                std::function<bool(tcp::GroupOp, Value)> shouldInlineConst)
+      : OpRewritePattern<GroupOp>(context),
+        shouldInlineConst_(shouldInlineConst) {}
+
+  LogicalResult matchAndRewrite(GroupOp groupOp,
                                 PatternRewriter &rewriter) const override {
     // Collect the values used in the given GroupOp. Those will be the inputs
     // to the IsolatedGroup op. The constants used in the GroupOp are collected
     // in a separate set and cloned into the body of the IsolatedGroupOp.
     llvm::SmallVector<Value> inputs;
     llvm::SmallDenseSet<Value> addedInputs;
     llvm::SmallDenseSet<Value> consts;
-    llvm::SmallDenseSet<Value> defs;
-    for (auto &op : groupOp.getBody().front()) {
-      for (auto operand : op.getOperands()) {
-        if (defs.find(operand) == defs.end()) {
+
+    groupOp->walk([&](Operation *op) {
+      for (auto operand : op->getOperands()) {
+        auto operandDefiningOp = operand.getDefiningOp();
+        if (!operandDefiningOp) {
+          operandDefiningOp = operand.getParentBlock()->getParentOp();
+        }
+        if (!groupOp->isProperAncestor(operandDefiningOp)) {
           if (operand.getDefiningOp() &&
-              operand.getDefiningOp()->hasTrait<OpTrait::ConstantLike>()) {
+              operand.getDefiningOp()->hasTrait<OpTrait::ConstantLike>() &&
+              shouldInlineConst_(groupOp, operand)) {
             consts.insert(operand);
           } else if (!addedInputs.contains(operand)) {
             inputs.push_back(operand);
             addedInputs.insert(operand);
           }
         }
       }
-      defs.insert(op.getResults().begin(), op.getResults().end());
-    }
+    });
 
-    auto isolatedGroupOp = rewriter.create<tcp::IsolatedGroupOp>(
+    auto isolatedGroupOp = rewriter.create<IsolatedGroupOp>(
```

Comment:
minor nit (feel free to ignore): consider keeping the `tcp::` scope to be consistent with other transforms.